ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/LedgerDB.hs

```diff
@@ -1,82 +1,1987 @@
+
+{- $snapshots
+
+  Snapshotting a ledger state means saving a copy of the in-memory part of the
+  ledger state serialized as a file on disk, as well as flushing differences on
+  the ledger tables between the last snapshotted ledger state and the one that
+  we are snapshotting now and making a copy of that resulting on-disk state.
+
+  == Startup
+
+  During initialisation, the goal is to construct an initial 'LedgerDB' where
+  the sequence of in-memory states is empty except for the ledger state at the
+  anchor, which has to correspond to the immutable tip, i.e., the block at the
+  tip of the Immutable DB.
+
+  Ideally, we can construct the initial 'LedgerDB' from a snapshot of the ledger
+  state that we wrote to disk. Remember that updating a ledger state with a
+  block is not invertible: we can apply a block to a ledger state, but we cannot
+  /unapply/ a block to a ledger state. This means the snapshot has to be at
+  least as old as the anchor. A snapshot matching the anchor can be used as is.
+  A snapshot older than the anchor can be used after reapplying the necessary
+  blocks. A snapshot newer than the anchor can /not/ be used, as we cannot
+  unapply blocks to get the ledger state corresponding to the anchor. This is
+  the reason why we only take snapshots of an immutable ledger state, i.e., of
+  the anchor of the 'DbChangelog' (or older).
+
+  On startup, the node will:
+
+  1. Find the latest snapshot which will be a directory identified by the slot
+     number of the snapshot:
+
+        > <cardano-node data dir>
+        > ├── volatile
+        > ├── immutable
+        > └── ledger
+        >     ├── <slotNumber1>
+        >     │   ├── tables
+        >     │   └── state
+        >     ├── <slotNumber2>
+        >     │   ├── tables
+        >     │   └── state
+        >     └── <slotNumber3>
+        >         ├── tables
+        >         └── state
+
+  2. Depending on the snapshots found, there are two possibilities:
+
+       - If there is no snapshot to load, create a new @'BackingStore'@ with the
+         contents of the Genesis ledger tables and finish.
+
+       - If there is a snapshot found, then deserialize (with @DecodeDisk@) the
+         @state@ file, which contains a serialization of the in-memory part of
+         the LedgerState, with empty tables (i.e. a @ExtLedgerState blk
+         EmptyMK@). If deserialization fails, delete this snapshot and start
+         again. If the snapshot is newer than the immutable tip, delete this
+         snapshot and start again.
+
+        In case we found an snapshot, we will overwrite (either literally
+        overwriting it or using some feature from the specific backend used) the
+        @BackingStore@ tables with the @tables@ file from said snapshot as it
+        was left in whatever state it was when the node shut down.
+
+  3. The deserialized ledger state will be then used as the anchor for the
```

Comment:
Are we using "anchor" with a different meaning here? Perhaps we should define these terms (or refresh what they mean) in this section.